### ISSUE TITLE
Feature Request: Support for Rust projects built with Bazel

### ISSUE THREAD

**Issue by Zintinio** [intellij-rust/intellij-rust#3730]
I've been working on integrating a Rust project with the [Bazel](https://bazel.build/) build system instead of using Cargo. So far I've managed to get the project building, but when trying to get the project loaded into CLion or Intellij, I get messages that "This file is not associated with a Cargo.toml", and jump to definition, highlighting, etc seems to break. I have raised an issue with the Bazel plugin authors about Rust support https://github.com/bazelbuild/intellij/issues/675 and have spent a little bit of time delving into how Intellij-Rust works to see if I can help out. 

Are the authors of this plugin interested in supporting non-Cargo build tools such as Bazel? If so, is this something I can help with? Roughly how much effort would this be? Thank you for your hard work, I know we're still a bit early in the days of Rust tooling.

**Comment by Zintinio** [comment]
Looking more closely at this, it seems that a minimal Rust application should be possible to support, one without a `Cargo.toml` specified. In the above example that I included, an application following the structure:

```
app/
├── BUILD
└── src
    ├── main.rs
    ├── one.rs
    └── two.rs
```

Should be analyzable by the IDE without the usage of Cargo. My intent would be to vendor dependencies, so from that perspective, it's all one giant module.

**Comment by vlad20012** [comment]
> Are the authors of this plugin interested in supporting non-Cargo build tools such as Bazel?

Yes, we're interested, but looks like it is very low-priority task and so It's very unlikely that we'll do it on our own :(

Looks like that in our plugin we should just provide some public API to give possibility to inject Rust project structure (some abstraction around our `CargoProjectsService`), and then this API should be used by the Bazel plugin or by some 3rd intermediate plugin (IJ Rust <-> Bazel bridge).

> Roughly how much effort would this be?

I dooon't sure. Looks like we are very tied to cargo project structure (it has some specific terms [project/workspace/package/target] and most of them aren't represented on rustc level), and I don't sure how it should looks like with other build tools. But initial point it `CargoProjectsService`, I guess.

> If so, is this something I can help with?

Well, if you don't mind Kotlin, you can try to hack around `CargoProjectsService`. I can mentor you from IJ Rust side, but to be honest, I know nothing about Bazel and its plugin =)

**Comment by Zintinio** [comment]
Thanks for the response. I will do some digging around, and see what I can learn. I think you have the right idea about exposing some kind of public interface that we can project structure into.

**Comment by petroseskinder** [comment]
Consider adding the support directly to https://github.com/bazelbuild/intellij.

The bazel intellij plugin works by annotating the Bazel build graph via [aspects](https://docs.bazel.build/versions/master/skylark/aspects.html). 

From a cursory glance through the codebase, it currently has support for go, cpp, dart, javascript, python, scala, kotlin and of course java. I don't see anything related to rust there. 

@jin would likely be the best person to reach out to as far as how to contribute rust support

**Comment by petroseskinder** [comment]
Oh I see you already filed https://github.com/bazelbuild/intellij/issues/675 :(

**Comment by TomaszWaszczyk** [comment]
I have it too:

![image](https://user-images.githubusercontent.com/1605705/93615581-e1353e00-f9d3-11ea-8b65-6f8bc8dfc48e.png)

**Comment by vlad20012** [comment]
c.c. #5594

**Comment by alexjpwalker** [comment]
I've opened a PR which integrates with the IntelliJ Bazel plugin to deliver Bazel support.
- #8491 

It depends on
- https://github.com/bazelbuild/intellij/pull/3222

I've uploaded an example project, and demo of the plugin's capabilities, at https://github.com/vaticle/bazel-intellij-rust-example

https://user-images.githubusercontent.com/14253102/151391801-a7551fcb-0322-4ed5-adc1-f11784a3b106.mov

However there are severe limitations to the PR currently (see the PR description for details). If someone familiar with Bazel is able to chime in, particularly on how to communicate Bazel sandbox path information from the Bazel plugin to the Rust plugin, that would be ideal.
